components/org.wso2.carbon.identity.tag.manager.core/src/main/java/org/wso2/carbon/identity/tag/manager/core/internal/impl/TagManagementServiceImpl.java

```diff
@@ -0,0 +1,271 @@
+import org.wso2.carbon.identity.tag.manager.model.ErrorMessage;
+import org.wso2.carbon.identity.tag.manager.model.Tag;
+import org.wso2.carbon.identity.tag.manager.model.TagAssociationsResult;
+import org.wso2.carbon.identity.tag.manager.model.TagListResult;
+import org.wso2.carbon.user.api.Tenant;
+import org.wso2.carbon.user.api.UserStoreException;
+
+import java.sql.SQLException;
+import java.util.List;
+
+/**
+ * Tag management service implementation class.
+ */
+public class TagManagementServiceImpl implements TagManagementService {
+
+    private static final Log LOG = LogFactory.getLog(TagManagementService.class);
+    private String tenantUuid = getTenantUuid();
+
+    public TagManagementServiceImpl() throws TagServiceException {
+
+    }
+
+    @Override
+    public String addTag(Tag tag) throws TagServiceException {
+
+        // Check if the tag name is valid.
+        if (!isTagNameValid(tag.getName())) {
+            throw new TagServiceClientException(ErrorMessage.ERROR_CODE_EMPTY_NAME.getCode(),
+                    ErrorMessage.ERROR_CODE_EMPTY_NAME.getMessage());
+        }
+        TagManagementDAO tagManagementDAO = DAOFactory.getInstance().getTagManagementDAO();
+        // Check if tag already existing.
+        if (tagManagementDAO.loadTag(tag.getName(), tag.getType(), tenantUuid) != null) {
+            throw new TagServiceClientException(ErrorMessage.ERROR_CODE_TAG_ALREADY_EXISTS.getCode(),
+                    String.format(ErrorMessage.ERROR_CODE_TAG_ALREADY_EXISTS.getMessage(), tag.getName(),
+                            tenantUuid));
+        }
+        String tagUuid = tagManagementDAO.storeTag(tag, tenantUuid);
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("Provided JSON request content is not in the valid format:");
+        }
+        return tagUuid;
+    }
+
+    @Override
+    public Tag getTagById(String tagUuid) throws TagServiceException {
+
+        Tag tag;
+        if (StringUtils.isBlank(tagUuid)) {
+            throw new TagServiceClientException(ErrorMessage.ERROR_CODE_EMPTY_TAG_ID.getCode(),
+                    ErrorMessage.ERROR_CODE_EMPTY_TAG_ID.getMessage());
+        }
+        TagManagementDAO tagManagementDAO = DAOFactory.getInstance().getTagManagementDAO();
+        tag = tagManagementDAO.loadTag(tagUuid, tenantUuid);
+        if (tag == null) {
+            throw new TagServiceClientException(
+                    ErrorMessage.ERROR_CODE_TAG_DATA_DOES_NOT_EXIST.getCode(),
+                    String.format(ErrorMessage.ERROR_CODE_TAG_DATA_DOES_NOT_EXIST.getMessage(), tagUuid));
+        }
+        return tag;
+    }
+
+    @Override
+    public List<String> getTagTypes(int limit, int offset, String sortBy, String sortOrder) throws TagServiceException {
+
+        List<String> tagList;
+        handleSorting(sortBy,sortOrder);
+        TagManagementDAO tagManagementDAO = DAOFactory.getInstance().getTagManagementDAO();
```

Comment:
We can validate the empty list here and negative offset

Reply:
done